pipit/trace.py

```diff
@@ -681,3 +681,65 @@ def calc_exc_time_in_bin(events):
         df.insert(0, "bin_end", edges[1:])
 
         return df
+
+    def detect_pattern(
+        self,
+        start_event,
+        iterations=None,
+        window_size=None,
+        process=0,
+        metric="time.exc",
+    ):
+        import stumpy
+
+        # count the number of enter events to
+        # determine the number of iterations if it's not
+        # given by the user.
+        if iterations is None:
+            iterations = len(
+                self.events[
+                    (self.events["Name"] == start_event)
+                    & (self.events["Event Type"] == "Enter")
+                    & (self.events["Process"] == process)
+                ]
+            )
+
+        # get the first enter and last leave of
+        # the given event. we will only investigate
+        # this portion of the data.
+        first_loop_enter = self.events[
+            (self.events["Name"] == start_event)
```

Comment:
this and line 716 seem like repeated filtering/computation since we already do it once on line 701. maybe we could store this "iterations" filter in a variable?

Reply:
Done.